### ISSUE TITLE
[Feature]: I want to contribute wandb proto scala3 project

### ISSUE THREAD

**Issue by mullerhai** [wandb/wandb#10137]
### Description

Hi,   perfect project ,I want to use your proto to generate scala3 wandb sdk project  for  scala3-pytorch  STorch

### Suggested Solution

dd

**Comment by mullerhai** [comment]
this is the sdk repo https://github.com/mullerhai/storch-wandb

**Comment by mullerhai** [comment]
only find  SystemMonitorServiceGrpc.bindService() api
 how to create wandbClient  with wandb proto scala3 generate code?

**Comment by mullerhai** [comment]
we think  we need more wandb api  in proto file  ,like inference  and table and so on . and  we need wandb server  api ,thanks, we need  your help

**Comment by mullerhai** [comment]
we just want to write wandb  code  like that 
```
package wandb_internal.wandb_base

import scala.sys.process._
import com.google.api.client.util.IOUtils
import java.io.File
import java.nio.file.{Files, Paths}
import com.google.protobuf.ByteString
import scalapb.GeneratedMessage

// 假设存在的 W&B Scala SDK 客户端类
class WandbClient {
  def init(project: String, id: String, resume: String): Unit = {
    // 初始化 W&B 运行
    println(s"Initializing W&B run for project $project with id $id and resume $resume")
  }

  def createArtifact(name: String, `type`: String): Artifact = {
    // 创建 W&B 工件
    new Artifact(name, `type`)
  }

  def logArtifact(artifact: Artifact): Unit = {
    // 记录 W&B 工件
    println(s"Logging artifact ${artifact.name} of type ${artifact.`type`}")
  }

  def finish(): Unit = {
    // 结束 W&B 运行
    println("Finishing W&B run")
  }
}

// 假设存在的 W&B 工件类
class Artifact(val name: String, val `type`: String) {
  def addDir(dirPath: String): Unit = {
    // 向工件添加目录
    println(s"Adding directory $dirPath to artifact $name")
  }

  def addFile(filePath: String): Unit = {
    // 向工件添加文件
    println(s"Adding file $filePath to artifact $name")
  }
}

object WandbDemo extends App {
  val wandb = new WandbClient()
  // 替换为实际的 run_id
  val runId = "your_run_id"
  // 初始化 W&B 运行
  wandb.init(project = "wandb_demo", id = runId, resume = "must")

  // 保存数据集
  val artiDataset = wandb.createArtifact("mnist", "dataset")
  artiDataset.addDir("mnist/")
  wandb.logArtifact(artiDataset)

  // 保存代码
  val artiCode = wandb.createArtifact("ipynb", "code")
  artiCode.addFile("./30分钟吃掉wandb可视化模型分析.ipynb")
  wandb.logArtifact(artiCode)

  // 替换为实际的检查点路径
  val ckptPath = "your_ckpt_path"
  // 保存模型
  val artiModel = wandb.createArtifact("cnn", "model")
  artiModel.addFile(ckptPath)
  wandb.logArtifact(artiModel)

  // 结束 W&B 运行
  wandb.finish()
}

```

**Comment by ghost** [comment]
Hi @mullerhai, thanks for reaching out! Would you be able to provide more context on why you want to do this and your full use case? Could you also tell me whether you're interested in using our public cloud or a local Docker image to do this? We'll speak with our SDK team to inquire about this following your response. Thanks in advance!

**Comment by mullerhai** [comment]
Hi ,now we use scala3 to train  pytorch model  use STorch,  need to tracing something loss aucurry , we have make Scala-tensorboard reality .  so we also want to use wandb in scala.

let me show you some Scala3 with  STorch with tensorBoard  code , then will inspire you  why we need wandb in Scala#


````
//import spire.random.Random

import basic.LstmNetApp.device
import basic.{LstmNet, NeuralNet}
import org.bytedeco.javacpp.{FloatPointer, PointerScope}
import org.bytedeco.pytorch.{OutputArchive, TensorExampleVectorIterator}
import torch.*
import torch.Device.{CPU, CUDA}
import torch.Layout.Strided
import torch.utils.data.dataset.ChunkSharedBatchDataset
import torch.nn.functional as F
import torch.nn.modules.HasParams
import torch.numpy.matrix.NDArray
import torch.numpy.serve.{Numpy, TorchNumpy}
import torch.optim.Adam

import java.nio.file.Paths
//import torchvision.datasets.FashionMNIST
import java.nio.file.Paths
//import scala.runtime.stdLibPatches.Predef.nn
import torch.internal.NativeConverters.{fromNative, toNative}
import scala.util.{Random, Using}
import torch.utils.tensorboard.SummaryWriter

object lstmGpu extends App {

  @main
  def main(): Unit =
//    System.setProperty( "org.bytedeco.javacpp.logger.debug" , "true")
    System.setProperty("org.bytedeco.javacpp.nopointergc", "true")
//    val path = "D:\\data\\git\\storch-numpy\\newNumpyArray.npy"
//    val np = new Numpy(path)
//    val npArray: NDArray[Double] = new Numpy(path).loadNDArray
//    println(npArray.getShape.mkString(","))
//    println(npArray.printArray())
//    val ten: Tensor[Float64] = Tensor(npArray, requires_grad = true, device = CPU)
//    println(ten)

    val logDir = "logs"
    val writer = new SummaryWriter(logDir, tfEventFilePath = "train.tfevents")
    val input_size = 28 * 28
    val hidden_size = 500
    val num_classes = 10
    val num_epochs = 100
    val batch_size = 100
    val learning_rate = 0.001f
    val dataPath = Paths.get("D:\\data\\FashionMNIST")
    val train_dataset = FashionMNIST(dataPath, train = true, download = true)
    val test_dataset = FashionMNIST(dataPath, train = false)
    println(s"torch.cuda.isAvailable ${torch.cuda.isAvailable}")
//    println(s"torch.cuda.device_count ${torch.cuda.}")
    val device = CUDA //if torch.cuda.isAvailable then CUDA else CPU
    val evalFeatures = test_dataset.features.to(device)
    val evalTargets = test_dataset.targets.to(device)
    val r = Random(seed = 0)

    def dataLoader: Iterator[(Tensor[Float32], Tensor[Int64])] =
      r.shuffle(train_dataset).grouped(8).map { batch =>
        val (features, targets) = batch.unzip
        (torch.stack(features).to(device), torch.stack(targets).to(device))
      }

    val model = LstmNet().to(device)
//    val model = NeuralNet[Float32](input_size, hidden_size, num_classes).to(device)
//    val model =   nn.Linear(input_size, num_classes).to(device) //LstmNet().to(device) //
    val criterion = nn.loss.CrossEntropyLoss().to(device)
    val optimizer = torch.optim.SGD(model.parameters(true), lr = learning_rate)

    import org.bytedeco.pytorch.{ChunkDatasetOptions, DataLoaderOptions, Example, ExampleIterator, ExampleStack, ExampleVector, ExampleVectorIterator, JavaDataset, JavaDistributedRandomTensorDataLoader, JavaDistributedSequentialTensorDataLoader, JavaRandomDataLoader, JavaRandomTensorDataLoader, JavaSequentialTensorDataLoader, JavaStatefulDataset, JavaStreamDataLoader, RandomSampler, SizeTArrayRef, SizeTOptional, TensorExample, TensorExampleIterator, TensorExampleStack, TensorExampleVector, AbstractTensor as Tensor, ChunkDataReader as CDR, ChunkDataset as CD, ChunkRandomDataLoader as CRDL, ChunkSharedBatchDataset as CSBD, DistributedRandomSampler as DRS, DistributedSequentialSampler as DSS, JavaStreamDataset as JSD, JavaTensorDataset as TD, StreamSampler as STS}
//    import torch.utils.data.DataLoaderOptions
    import torch.utils.data.dataloader.*
    import torch.utils.data.datareader.{ChunkDataReader, ChunkTensorDataReader, ExampleVectorReader, TensorExampleVectorReader}
    import torch.utils.data.dataset.*
    import torch.utils.data.dataset.java.{StatefulDataset, StatefulTensorDataset, StreamDataset, StreamTensorDataset, TensorDataset, JavaDataset as JD}
    import torch.utils.data.sampler.{DistributedRandomSampler, DistributedSequentialSampler, StreamSampler, RandomSampler as RS, SequentialSampler as SS}

    def exampleVectorToExample(exVec: ExampleVector): Example = {
      val example = new Example(exVec.get(0).data(), exVec.get(0).target())
      example
    }


    val exampleSeq = train_dataset.map(x => new Example(x._1.native, x._2.native))
    //  val ex1 = new Example(mnistTrain.features.native ,mnistTrain.targets.native)
    val exampleVector = new ExampleVector(exampleSeq *)
    val reader = new ChunkDataReader()
    reader(exampleVector)
    val prefetch_count = 1
    //  val ds = new ChunkSharedBatchDataset(new ChunkDataset(reader, new RandomSampler(exampleSeq.size), new RandomSampler(exampleSeq.size), new ChunkDatasetOptions(prefetch_count, batch_size))).map(new ExampleStack)
    //  val ds  = new ChunkSharedTensorBatchDataset(new ChunkTensorDataset(reader,new RS(exampleTensorSeq.size),new ChunkDatasetOptions(prefetch_count, batch_size))).map(new TensorExampleStack)
    val ds = new ChunkSharedBatchDataset(
      new ChunkDataset(
        reader,
        new RandomSampler(exampleSeq.size),
        new RandomSampler(exampleSeq.size),
        new ChunkDatasetOptions(prefetch_count, batch_size)
      )
    ).map(new ExampleStack)

    val opts = new DataLoaderOptions(100)
    //  opts.workers.put(5)
    opts.batch_size.put(100)
    //  opts.enforce_ordering.put(true)
    //  opts.drop_last.put(false)
    val data_loader = new ChunkRandomDataLoader(ds, opts)
    val total_step = train_dataset.length // 2000 //data_loader //
    (1 to num_epochs).foreach(epoch => {
      var it: ExampleIterator = data_loader.begin
      var batchIndex = 0
      println("coming in for loop")
      while (!it.equals(data_loader.end)) {
        Using.resource(new PointerScope()) { p =>
          val batch = it.access
          optimizer.zeroGrad()
//          println(s"batch comming in ${batchIndex}")
//          val trainDataTensor = fromNative(batch.data()).to(device)

//          val weight = fromNative(batch.data()view(-1, 28))
//          import torch.numpy.enums.DType as NumpyDType
//          val npDtype = NumpyDType.Float32
//          val shape = weight.sizes().map(_.toLong).toArray

//          val floatData = new Array[Float](weight.native.numel().toInt)
//          val  ft = new FloatPointer(weight.native.data_ptr_float())
//          ft.get(floatData)
////          weight.native.data_ptr().asInstanceOf[FloatPointer].get(floatData)
//          floatData.asInstanceOf[Array[DTypeToScala[Float32]]]
////          println(s"weight shape ${floatData.mkString(",")}")
//          val ndArray2 = new NDArray[DTypeToScala[Float32]](
//            data = floatData,
//            shape = shape.map(_.toInt),
//            ndim = shape.size,
//            dType = npDtype,
//          ).reshape(shape.map(_.toInt) *)
//          println(s"weight shape ${ndArray2.printArray()}")

//          println(s"weight shape ${weight.shape.mkString(",")}")
//          val npArr = torch.toNDArray(weight.view(-1,2,2,28,28))

//          val npArr = torch.toNDArray(weight.view(-1,2,2, 28, 28))
//          println(s"NDArray shape ${npArr.getShape.mkString(",")}")
//          println(npArr.printArray())
//          val requires_grad = true
//          val ndArray = npArr.getArray

          import torch.Tensor as TorchTensor
//          val tensor: TorchTensor[Float32] = ndArray match {
//            case singleDim: Array[Float] =>
//              val dataSeq = singleDim.toSeq
//              TorchTensor(dataSeq, Strided, device, requires_grad)
//            case twoDim: Array[Array[Float]] =>
//              val dataSeq = twoDim.map((arr: Array[Float]) => arr.toSeq).toSeq
//              TorchTensor(dataSeq, Strided, device, requires_grad)
//            case threeDim: Array[Array[Array[Float]]] =>
//              val dataSeq = threeDim.map((arr: Array[Array[Float]]) => arr.map(_.toSeq).toSeq).toSeq
//              TorchTensor(dataSeq, Strided, device, requires_grad)
//            case fourDim: Array[Array[Array[Array[Float]]]] =>
//              val dataSeq = fourDim.map((arr: Array[Array[Array[Float]]]) => arr.map(_.map(_.toSeq).toSeq).toSeq).toSeq
//              TorchTensor(dataSeq, Strided, device, requires_grad)
//            case fiveDim: Array[Array[Array[Array[Array[Float]]]]] =>
//              val dataSeq = fiveDim.map((arr: Array[Array[Array[Array[Float]]]]) => arr.map(_.map(_.map(_.toSeq).toSeq).toSeq).toSeq).toSeq
//              TorchTensor(dataSeq, Strided, device, requires_grad)
//            case _ => throw new IllegalArgumentException("Unsupported array dimension")
//          }
//          println(tensor)


//          tensor
//          val npArr2 :NDArray[Double] = TorchNumpy.rand(Array(3,6),NumpyDType.Float32)
//          val tensor2 = torch.NDArrayToTensor[Double](npArr2,false,CPU)
//          val tensor3 = torch.fromNDArray[Double](npArr2,false,CPU)
//          println(tensor2)
//          println(tensor3)

          //          println(s"ze: ${ze(0)} 1: ${ze(1)} 2 : ${ze(2)} ${ze(3)}")
          val prediction = model(fromNative(batch.data().view(-1, 28 * 28)).reshape(-1, 28 , 28).to(device))
          val loss = criterion(prediction, fromNative(batch.target()).to(device))
          writer.addScalar("Training/loss", loss.item, (epoch-1)*600+batchIndex)
          loss.backward()
          optimizer.step()
          it = it.increment
//          val accuracy =
//            (prediction.argmax(dim = 1).eq(fromNative(batch.target()).to(device)).sum / train_dataset.length).item
//
//          println(
//            f"Epoch: $epoch | Batch: $batchIndex%4d | Training loss: ${loss.item}%.4f | Training loss: ${loss.item}%.4f | Eval accuracy: $accuracy%.4f"
//          )
//          println(s"train Loss grad_fn: ${loss.grad_fn()}")
//          println(s"train out grad_fn: ${prediction.grad_fn()}")
          batchIndex += 1
          if batchIndex % 200 == 0 then
            // run evaluation
            torch.noGrad {
              val correct = 0
              val total = 0
              println("coming eval...")
              val predictions = model(evalFeatures.reshape(-1, 28 ,28))
              println(s"predictions : ${predictions} \n")
              val evalLoss = criterion(predictions, evalTargets)
              println(s"evalLoss : ${evalLoss.item} \n")
//              val featuresData = new Array[Float](1000)
//              val fp4 = new FloatPointer(predictions.native.data_ptr_float())
//              fp4.get(featuresData)
//              println(s"\n ffff size ${featuresData.size} shape ${
//                evalFeatures.shape
//                  .mkString(", ")
//              }a data ${featuresData.mkString(" ")}")
              println(s"predictions : ${predictions} \n")
              writer.addScalar("Evaling/loss", evalLoss.item, (epoch-1)*600+batchIndex)
              //              println(s"loss grad_fn: ${evalLoss.grad_fn()}")
              val accuracy =
                (predictions.argmax(dim = 1).eq(evalTargets).sum / test_dataset.length).item
              println(
                f"Epoch: $epoch | Batch: $batchIndex%4d | Training loss: ${loss.item}%.4f | Eval loss: ${evalLoss.item}%.4f | Eval accuracy: $accuracy%.4f"
              )
            }

          //        it = it.increment

        }
      }

    })
    writer.close()
    println(s"model ${model.modules.toSeq.mkString(" \n")}")
    println(s"model ${model.summarize}")

  //    Using.resource(torch.noGrad()){
  //      
  //    }
}

```

**Comment by ghost** [comment]
That great! All of our proto definitions are [here](https://github.com/wandb/wandb/tree/main/wandb/proto). There's also an example [here](https://github.com/wandb/wandb/tree/main/experimental/client-csharp) of how to build a client for our service (this is for C#).

However, please note that the wandb core interface will evolve over time as we make changes without announcing or documenting them as these are internal implementation details. Please bear this in mind with your project!

**Comment by mullerhai** [comment]
> That great! All of our proto definitions are [here](https://github.com/wandb/wandb/tree/main/wandb/proto). There's also an example [here](https://github.com/wandb/wandb/tree/main/experimental/client-csharp) of how to build a client for our service (this is for C#).
> 
> However, please note that the wandb core interface will evolve over time as we make changes without announcing or documenting them as these are internal implementation details. Please bear this in mind with your project!

the C #  demo , I think  it is not ok ,I need wandb   provide more proto  for build other language sdk

**Comment by ghost** [comment]
Thanks, just so I am clear, would you mind expanding on "wandb provide more proto for building another language SDK"?
